python/cudf/cudf/core/column/categorical.py

```diff
@@ -501,24 +501,30 @@ def find_and_replace(self, to_replace, replacement, all_nan):
         """
         replaced = column.as_column(self.cat().codes)
 
+        to_replace_col, replacement_col = [], []
+        new_cats = cudf.Series(self.dtype.categories)
+        for old_val, new_val in zip(to_replace, replacement):
+            if new_val not in self.dtype.categories:
+                new_cats = new_cats.replace(old_val, new_val)
+            else:
+                to_replace_col.append(self._encode(old_val))
+                replacement_col.append(self._encode(new_val))
+
         to_replace_col = column.as_column(
-            np.asarray(
-                [self._encode(val) for val in to_replace], dtype=replaced.dtype
-            )
+            np.array(to_replace_col, dtype=replaced.dtype)
         )
         replacement_col = column.as_column(
-            np.asarray(
-                [self._encode(val) for val in replacement],
-                dtype=replaced.dtype,
-            )
+            np.array(replacement_col, dtype=replaced.dtype)
         )
 
+        replaced = column.as_column(self.cat().codes)
+
         output = libcudf.replace.replace(
             replaced, to_replace_col, replacement_col
         )
 
         return column.build_categorical_column(
-            categories=self.dtype.categories,
+            categories=new_cats,
```

Comment:
This fixes the current issue where we error if we try and replace with a category that doesn't exist yet. However we still differ from Pandas in terms of the result we get if we replace with an *existing category*. Previously, pandas would cast the entire column to object in this case and return the corresponding object column with the expected values replaced. However going forward, pandas returns a new categorical column with a new categorical dtype corresponding to the remaining categories in the list. If the categories are ordered, the new datatype will also be ordered and the replaced values will be in the same "spot" in the ordering as the old ones. 

This is one of the reasons I'm against retaining the old category in the replaced column's datatype - perhaps it wouldn't be a bad choice in the unordered case, but in the ordered case we end up with two categories that both only make sense in the same spot in the ordering.

Reply:
That is the behavior in pandas as far as I can tell, however I do agree that it feels wrong in that scenario - why should replacing `gold` with `bronze` in the data change my datatype such that `bronze < silver`? My concern is that if we go the other way, and keep `gold` in the datatype, I'm not sure what to do if we need to do something like `replace('gold', 'platinum')`. Do we get `[bronze < silver < gold < platinum]` or do we get `[bronze < silver < platinum < gold]` ?

Reply:
I'd think we should just throw from being ambiguous at that point, but either way lets just follow Pandas for now and we can change our behavior if desires arise.